### ISSUE TITLE
Codesigning an app bundle gives "initfsencoding: unable to load the file system codec" error on launch

### ISSUE THREAD

**Issue by stefanfis** [ronaldoussoren/py2app#291]
tl;dr

Code signing the launch binary in a previously working app bundle gives the error "initfsencoding: unable to load the file system codec".

---

I'm packaging a wxpython application using py2app:

    python setup.py py2app --packages="wx"

This step works well and creates an app bundle that I can launch successfully.

Then I try to `codesign` the app bundle. First step is to sign all the libs:

    find "myapp.app" -iname '*.so' -or -iname '*.dylib' |
        while read libfile; do
            codesign --sign "${IDENTITY}" "${libfile}" --options=runtime;
        done;
    codesign --sign "${IDENTITY}" --options=runtime "myapp.app/Contents/MacOS/python"

These steps again work as well and I can still launch the application, which is not yet fully signed as the `myapp` binary in `Contents/MacOS/` hasn't been signed so far.

Finally I sign the `myapp` binary:

    codesign --sign "${IDENTITY}" --options=runtime "myapp.app/Contents/MacOS/myapp"

Right after this last codesign, `myapp.app` stops working. Opening it with `./myapp.app/Contents/MacOS/myapp` gives this error:

    Fatal Python error: initfsencoding: unable to load the file system codec
    Traceback (most recent call last):
      File "/Users/xy/Dropbox/Projekte/myapp/dist/myapp.app/Contents/Resources/lib/python3.7/encodings/__init__.py", line 31, in <module>
    zipimport.ZipImportError: can't decompress data; zlib not available
    Abort trap: 6

The file `encodings/__init__.py` where the error occurs is the 'Standard "encodings" Package' and line 31 reads `import codecs`. (`codesign` worked without a problem again in this last step. I can even successfully continue to notify the app bundle, but it just won't launch.)

Looking for zlib library yields:

	$ find dist/myapp.app/ -iname "*zlib*"
	dist/myapp.app//Contents/Resources/zlib.cpython-37m-darwin.so
	dist/myapp.app//Contents/Resources/lib/python3.7/lib-dynload/zlib.so

I tried to include several "packages" and "includes" in setup.py that I thought could help, but the error is still there:

    OPTIONS = {
        'packages': ['wx', 'encodings'],
        'includes': ['zlib', 'codecs']
    }

---

Used software:
macOS Catalina 10.15.4
Python: 3.7.7
py2app: 0.21
wx: 4.0.4 osx-cocoa (phoenix) wxWidgets 3.0.5
Everything installed with `conda` in an environment.

**Comment by hasii2011** [comment]
Any other comments on this.  I am getting ready to plunk down $99 on the Apple Developer program so I can get code signing key?

**Comment by hamishmb** [comment]
I use codesign --deep with an ad hoc key at the moment, and I don't have this issue.

However, ad hoc keys don't do anything other than providing a checksum, so that may be why. --deep may also not actually work quite right for systems that require it eg Apple Silicon native builds. Currently, I'm using Rosetta 2 as wxPython doesn't have an Apple Silicon build out yet.
